provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassloadingCache.kt

```diff
@@ -0,0 +1,115 @@
+ * limitations under the License.
+ */
+package org.gradle.kotlin.dsl.provider
+
+import org.gradle.api.internal.initialization.ClassLoaderScope
+
+import org.gradle.cache.internal.CrossBuildInMemoryCache
+import org.gradle.cache.internal.CrossBuildInMemoryCacheFactory
+
+import org.gradle.internal.Cast
+import org.gradle.internal.classpath.DefaultClassPath
+import org.gradle.internal.hash.HashCode
+
+import org.gradle.kotlin.dsl.support.loggerFor
+
+import java.io.File
+import java.lang.ref.WeakReference
+import javax.inject.Inject
+
+
+internal
+data class LoadedScriptClass<out T : Any>(
+    val compiledScript: CompiledScript<T>,
+    val scriptClass: Class<*>)
+
+
+internal
+class KotlinScriptClassloadingCache @Inject constructor(cacheFactory: CrossBuildInMemoryCacheFactory) {
+
+    private
+    val logger = loggerFor<KotlinScriptPluginFactory>()
+
+    private
+    val cache: CrossBuildInMemoryCache<ScriptCacheKey, LoadedScriptClass<*>> = cacheFactory.newCache()
+
+    fun <T : Any> loadScriptClass(
+        scriptBlock: ScriptBlock<T>,
+        parentClassLoader: ClassLoader,
+        scopeFactory: () -> ClassLoaderScope,
+        compilation: (ScriptBlock<T>) -> CompiledScript<T>
+    ): LoadedScriptClass<T> {
+
+        val key = ScriptCacheKey(scriptBlock.scriptTemplate.qualifiedName!!, scriptBlock.sourceHash, parentClassLoader)
+        val cached = cache.get(key)
+        if (cached != null) {
+            return Cast.uncheckedCast<LoadedScriptClass<T>>(cached)
+        }
+
+        val compiledScript = compilation(scriptBlock)
+
+        logger.debug("Loading {} from {}", scriptBlock.scriptTemplate.simpleName, scriptBlock.displayName)
+        val scriptClass = classFrom(compiledScript, scopeFactory())
+        return LoadedScriptClass(compiledScript, scriptClass).also {
+            cache.put(key, it)
+        }
+    }
+
+    private
+    fun classFrom(compiledScript: CompiledScript<*>, scope: ClassLoaderScope): Class<*> =
+        classLoaderFor(compiledScript.location, scope)
+            .loadClass(compiledScript.className)
+
+    private
+    fun classLoaderFor(location: File, scope: ClassLoaderScope) =
+        scope
+            .local(DefaultClassPath(location))
+            .lock()
+            .localClassLoader
+}
+
+private
+class ScriptCacheKey(
+    private val templateId: String,
+    private val sourceHash: HashCode,
+    parentClassLoader: ClassLoader) {
+
+    private
+    val parentClassLoader: WeakReference<ClassLoader> = WeakReference(parentClassLoader)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
+            return true
+        }
+        if (other == null || this::class != other::class) {
+            return false
+        }
+        val key = other as ScriptCacheKey
+        return (parentClassLoader.get() != null && key.parentClassLoader.get() != null
+            && parentClassLoader.get() == key.parentClassLoader.get()
```

Comment:
The GC can't clear this reference between these two checks? If it can, you should store the loaders as local variables before comparing them.